dpp/src/dash_platform_protocol.rs

```diff
@@ -0,0 +1,159 @@
+    public_key: JSONSchema,
+}
+
+impl<'a> IdentitySchemas {
+    pub fn new(
+        identity_json: &'a serde_json::Value,
+        public_key_json: &'a serde_json::Value,
+    ) -> Result<Self, ValidationError<'a>> {
+        Ok(Self {
+            public_key: JSONSchema::compile(public_key_json)?,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct SchemaCompilationError {
+    json_schemas: JsonSchemas,
+}
+
+impl SchemaCompilationError {
+    pub fn new(json_schemas: JsonSchemas) -> Self {
+        Self { json_schemas }
+    }
+
+    pub fn error(&self) -> Option<ValidationError> {
+        self.json_schemas.compilation_error()
+    }
+}
+
+#[derive(Debug)]
+pub struct JsonSchemas {
+    schema_jsons: SchemaJsons,
+    identity: Option<IdentitySchemas>,
+}
+
+impl JsonSchemas {
+    pub fn new(schema_jsons: SchemaJsons) -> Self {
+        Self {
+            schema_jsons,
+            identity: None,
+        }
+    }
+
+    pub fn compile(&mut self) -> Result<(), ValidationError> {
+        match IdentitySchemas::new(
+            &self.schema_jsons.identity.identity_json,
+            &self.schema_jsons.identity.public_key_json,
+        ) {
+            Ok(identity_schemas) => {
+                self.identity = Some(identity_schemas);
+                Ok(())
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    pub fn compilation_error(&self) -> Option<ValidationError> {
+        match IdentitySchemas::new(
+            &self.schema_jsons.identity.identity_json,
+            &self.schema_jsons.identity.public_key_json,
+        ) {
+            Ok(_) => None,
+            Err(err) => Some(err),
+        }
+    }
+}
+
+pub struct DashPlatformProtocol {
+    protocol_version_validator: ProtocolVersionValidator,
+    json_schemas: JsonSchemas,
+
+    pub identities: IdentityFacade,
+}
+
+pub struct DPPOPtions {
+    current_protocol_version: Option<u32>,
+}
+
+impl Default for DPPOPtions {
+    fn default() -> Self {
+        Self {
+            current_protocol_version: None,
+        }
+    }
+}
+
+impl DashPlatformProtocol {
+    pub fn new(options: DPPOPtions) -> Result<Self, DashPlatformProtocolInitError> {
+        let schema_jsons = SchemaJsons::new()?;
+        let mut json_schemas = JsonSchemas::new(schema_jsons);
+        json_schemas.compile()?;
+
+        let current_protocol_version = options.current_protocol_version.unwrap_or(LATEST_VERSION);
+
+        let protocol_version_validator = ProtocolVersionValidator::new(
+            current_protocol_version,
+            LATEST_VERSION,
+            COMPATIBILITY_MAP.clone(),
+        );
+        let rc = Arc::new(protocol_version_validator);
+
+        Ok(Self {
+            json_schemas,
+            protocol_version_validator: ProtocolVersionValidator::new(
+                current_protocol_version,
+                LATEST_VERSION,
+                COMPATIBILITY_MAP.clone(),
+            ),
+            identities: IdentityFacade::new(rc.clone())?,
```

Comment:
probably, cloning here, is not necessary.

Reply:
The reason for that is that the protocol version validators need to be everywhere - in all future validators as well. There are multiple ways to deal with it:
 1. Clone the protocol version validator and let all other validators own their own copy of it. Not a big deal, since the version validator is tiny, it has only three parameters to it - two u32 and one compatibility map, which is as of now just two u32s as well. The map can grow over time, but the protocol version is not something we’re going to update often. The problem with every validator owning its own version validator is though that it would be technically possible to have different configurations of the version validator in the same codebase if someone doesn’t watch closely what they’re doing.
 2. Pass a reference to the same instance of the version validator from the DashPlatformProtocol. That would be the best solution technically, but then you can’t just use `DashPlatformProtocol::new`, because it would be impossible to initialize the version validator inside the new and pass references to it to all other validators. That can be done by adding an init function, which works as well, and what currently is done in JS, so no biggie, I guess, just looks a little bit strange that the instance created with `::new` can not be used right away

So to bypass the cloning and configuration issue of the option 1 and the strange ergonomics of the option 2, I’ve added an Arc around the version validator

(Answering the question in your comment - all the future validators will need the access to the version validator, hence the `rc.clone`)

It is, of course, still possible to mess with the configuration that way too, just a little bit harder, and the interface clearly implies that the version validator has to be shared with other components and not owned